src/build/rewrite-webcomponents-loader.ts

```diff
@@ -49,24 +78,73 @@ export class UseES5WebcomponentsLoader extends stream.Transform {
       });
     }
-    if (!/webcomponents-loader\.js/.test(contents)) {
+
+    if (!webcomponentsLoaderRegex.test(contents)) {
       callback(null, file);
       return;
     }
+
     const parsed = parse5.parse(contents);
-    const script = dom5.nodeWalk(parsed, scriptIncludeWebcomponentsLoader);
+    const script = dom5.nodeWalk(parsed, webcomponentsLoaderMatcher);
     if (!script) {
       callback(null, file);
       return;
     }
 
-    const scriptPath = dom5.getAttribute(script, 'src')!;
-    const scriptUrl = url.resolve(scriptPath, 'custom-elements-es5-adapter.js');
-    const es5AdapterScript = parse5.parseFragment(`<script src="${scriptUrl}"></script>`);
-    dom5.insertBefore(script.parentNode!, script, es5AdapterScript);
-
+    // Collect important dom references & create fragment for injection
     const correctedFile = file.clone();
+    const bodyElement = dom5.query(parsed, bodyMatcher)!;
+    const headElement = dom5.query(parsed, headMatcher)!;
+    const loaderScriptUrl = dom5.getAttribute(script, 'src')!;
+    const adapterScriptUrl =
+        url.resolve(loaderScriptUrl, 'custom-elements-es5-adapter.js');
+    const loaderMovedComment = dom5.constructors.comment(
+        ' AUTO-GENERATED BY JS-COMPILE: "webcomponents-loader.js" and ' +
+        'related resources have been moved to <body> so that ' +
+        '"custom-elements-es5-adapter.js" polyfill could be properly ' +
+        'injected into this file. ');
+    const es5AdapterScript = parse5.parseFragment(`
+<div id="autogenerated-ce-es5-shim">
+  <script type="text/javascript">
+    // This prevents custom-elements-es5-adapter.js from parsing & running
+    // on browsers without native support for ES2015+ & custom elements.
+    if (!window.customElements) {
+      var ceShimContainer = document.querySelector('#autogenerated-ce-es5-shim');
+      ceShimContainer.parentElement.removeChild(ceShimContainer);
+    }
+  </script>
+  <script type="text/javascript" src="${adapterScriptUrl}"></script>
+</div>
+`);
+
+    // If script is in the body, just insert the es5 adapter script before it
+    if (dom5.nodeWalkAncestors(script, bodyMatcher)) {
+      dom5.insertBefore(bodyElement, script, es5AdapterScript);
+      correctedFile.contents = new Buffer(parse5.serialize(parsed), 'utf-8');
+      callback(null, correctedFile);
+      return;
+    }
+
+    // otherwise we need to move the webcomponents-loader/webcomponents-lite
+    // loader down to the body so that the es5 adaperter script shim will work
+    const scriptSiblings = dom5.queryAll(
+        script.parentNode!,
+        dom5.predicates.OR(scriptMatcher, styleMatcher, linkMatcher));
+    const scriptSiblingsFollowing =
+        scriptSiblings.splice(scriptSiblings.indexOf(script));
+    dom5.insertBefore(headElement, script, loaderMovedComment);
+    if (!bodyElement.childNodes) {
+      bodyElement.childNodes = [];
+    }
+    for (let i = 1, len = scriptSiblingsFollowing.length; i <= len; i++) {
```

Comment:
I have a phobia of complex for loops. Indexing is so finicky, and it's so easy to get it slightly wrong. It'd also be worth a comment about why we're doing this backwards. So I'd recommend instead:

```typescript
// Elements are pushed into the top of body, so need to insert them in reverse order.
scriptSiblingsFollowing.reverse();
for (const releventScript of scriptSiblingsFollowing) {
  dom5.insertBefore(
          bodyElement, bodyElement.childNodes[0] || null, releventScript);
}
```